lib/LedgerSMB/Upgrade_Tests.pm

```diff
@@ -1069,22 +1240,37 @@ push @tests, __PACKAGE__->new(
     );
 
 push @tests, __PACKAGE__->new(
-    test_query => q{select concat(ac.trans_id,'-',ac.id) as id,
+    test_query => q{select distinct concat(ac.trans_id,'-',ac.id) as id, c.accno,
                           ap.transdate, ap.datepaid,
-                          ac.cleared-ac.transdate as delay, ap.amount,v.name,
-                          ac.transdate,ac.cleared
+                          ac.cleared-ac.transdate as delay, cast(ap.amount as numeric(9,2)), v.name,
+                          ac.transdate,ac.cleared,
+                          case when ac.cleared-ac.transdate > 0 then 'warning'
+                               when ac.cleared-ac.transdate < 0
+                             or ap.datepaid is null
+                             or ac.cleared is null and ap.datepaid is not null
+                                                            then 'error'
+                               else                                  'good'
+                          end as validity
                   from ap
                   join acc_trans ac on ap.id=ac.trans_id
                   left join vendor v on v.id=ap.vendor_id
-                  where ((ac.cleared-ac.transdate > 150 or ac.cleared-ac.transdate < 0)
-                         or ac.cleared < ap.datepaid and ac.id = (select max(id) from acc_trans where ap.id=acc_trans.trans_id))
+                  left join chart c on c.id=ac.chart_id
+                  where ((ac.cleared-ac.transdate > 60 or ac.cleared-ac.transdate < 0)
+                         or (ac.cleared is null and ap.datepaid is not null and
+                            ( select count(cleared) from acc_trans
+                              where trans_id = ap.id ) = 0 or
+                             ac.cleared < ap.datepaid or ac.cleared is not null and ap.datepaid is null)
+                             and ac.id = (select max(id) from acc_trans where ap.id=acc_trans.trans_id))
+                    and c.link ~ 'paid'
                     and ac.id > 0
-                  order by ac.cleared,id, ac.transdate, ap.datepaid},
+                  order by c.accno, ap.transdate, id, ac.cleared, ap.datepaid},
   display_name => marktext('Invalid or suspect cleared delays'),
           name => 'invalid_cleared_dates',
-  display_cols => ['name', 'id', 'datepaid', 'transdate', 'cleared', 'delay', 'amount'],
+  display_cols => ['name', 'accno', 'id', 'transdate', 'datepaid', 'cleared', 'delay', 'amount'],
+     alignment => {accno => 'center', transdate => 'center', datepaid => 'center',
+                    cleared => 'center', delay => 'right', amount => 'right'},
  instructions => marktext(
-                   'Suspect or invalid cleared delays have been detected. Please review the dates in the original application'),
+                   'Cleared dates much later than the transaction dates are suspect. Cleared dates before the transaction dates are invalid. Please review the dates in the original application'),
```

Comment:
I think this change falls outside the scope of "disable migration when there's no AR or AP account", right?

If/When we decide to merge both functionalities, I think we want them in separate PRs.